src/PreparedWrites.h

```diff
@@ -0,0 +1,227 @@
+
+    /// Copy of the PreparedOp header.
+    Header header;
+
+    /// Pointer to the array of #WireFormat::TxParticipant which contains
+    /// all information of objects participating transaction.
+    WireFormat::TxParticipant* participants;
+
+    /// Object to be written during COMMIT phase.
+    /// Its value is empty if OpType is READ or REMOVE; it only contains
+    /// key information in that case.
+    Object object;
+
+    void assembleForLog(Buffer& buffer);
+    bool checkIntegrity();
+    uint32_t computeChecksum();
+
+    DISALLOW_COPY_AND_ASSIGN(PreparedOp);
+};
+
+class PreparedOpTombstone {
+  public:
+    PreparedOpTombstone(PreparedOp& op, uint64_t segmentId);
+    explicit PreparedOpTombstone(Buffer& buffer, uint32_t offset = 0);
+
+    void assembleForLog(Buffer& buffer);
+    bool checkIntegrity();
+    uint32_t computeChecksum();
+
+    /**
+     * This data structure defines the format of an preparedOp's tombstone
+     * in a master server's log.
+     */
+    class Header {
+      public:
+        Header(uint64_t tableId,
+               KeyHash keyHash,
+               uint64_t leaseId,
+               uint64_t rpcId,
+               uint64_t segmentId)
+            : tableId(tableId),
+              keyHash(keyHash),
+              leaseId(leaseId),
+              rpcId(rpcId),
+              segmentId(segmentId),
+              checksum(0)
+        {
+        }
+
+        /// TableId for log distribution during recovery.
+        uint64_t tableId;
+
+        /// KeyHash for log distribution during recovery.
+        KeyHash keyHash;
+
+        /// leaseId given for the prepare that this tombstone is for.
+        /// A (leaseId, rpcId) tuple uniquely identifies a preparedOp log entry.
+        uint64_t leaseId;
+
+        /// rpcId given for the prepare that this tombstone is for.
+        /// A (leaseId, rpcId) tuple uniquely identifies a preparedOp log entry.
+        uint64_t rpcId;
+
+        /// The log segment that the dead preparedOp this tombstone refers to
+        /// was in. Once this segment is no longer in the system, this tombstone
+        /// is no longer necessary and may be garbage collected.
+        uint64_t segmentId;
+
+        /// CRC32C checksum covering everything but this field, including the
+        /// key.
+        uint32_t checksum;
+    } __attribute__((__packed__));
+    static_assert(sizeof(Header) == 44,
+        "Unexpected serialized ObjectTombstone size");
+
+    /// Copy of the tombstone header that is in, or will be written to, the log.
+    Header header;
+
+    /// If a tombstone is being read from a serialized copy (for instance, from
+    /// the log), this will point to the buffer that refers to the entire
+    /// tombstone. This is NULL for a new tombstone that is being constructed.
+    Buffer* tombstoneBuffer;
+
+    DISALLOW_COPY_AND_ASSIGN(PreparedOpTombstone);
+};
+
+/**
+ * A table for all staged PreparedOp. Decision RPC handler fetches preparedOp
```

Comment:
What does "all staged PreparedOp" mean? Is this all PreparedOps for a particular transaction, or all PreparedOps for all currently executing transactions on a server? Be precise.